src/aspire/denoising/adaptive_support.py

```diff
@@ -0,0 +1,107 @@
+import logging
+
+import numpy as np
+
+from aspire.image import Image
+from aspire.numeric import fft
+from aspire.utils.coor_trans import grid_2d
+
+logger = logging.getLogger(__name__)
+
+
+def adaptive_support(images, energy_threshold=0.99):
+    """
+    Determine size of the compact support in both real and Fourier Space.
+
+    Images are scaled in real space so that the noise variance in both
+    real and Fourier domains is similar.
+
+    Returns c_limit (support in Fourier space), R_limit (support in real space).
+
+    Fourier c_limit is scaled in range [0, 0.5].
+    R_limit is in pixels [0, Image.res/2].
+
+    :param images: Input `Image` stack
+    :param energy_threshold: [0, 1] threshold limit
+    :return: (c_limit, R_limit)
+    """
+
+    # Sanity Check Input is Image stack
+    if not isinstance(images, Image):
+        raise RuntimeError("adaptive_support expects Image instance")
+
+    # Sanity Check Threshold is in range
+    if energy_threshold <= 0 or energy_threshold > 1:
+        raise ValueError(
+            f"Given energy_threshold {energy_threshold} outside sane range [0,1]"
+        )
+
+    L = images.res
+    N = L // 2
+    P = images.n_images
+
+    r = grid_2d(L, shifted=False, normalized=False, dtype=images.dtype)["r"]
+
+    # Transform to Fourier space
+    imgf = fft.centered_fft2(images.asnumpy())
+
+    # demean
+    img = images.asnumpy() * L
```

Comment:
This is odd.

Reply:
I thought so too, but at the time I assumed the original code was more mature than it turned out.  I'll try to back out why that might be there.

Reply:
The only thing I can think of is rescaling the variance estimate (effectively multiplying it by `L ** 2`), but it's not clear to me why this is needed.

Reply:
It's not needed.  I think it was an artifact in the old code relating to expectations of the input data.  The old code if it was ever run was getting a bespoke array from class averaging code. I've removed.